File: MATTHEW/leetcode/AIcoding/1.Gemini-2.5/javascript_solutions/2312-selling-pieces-of-wood.js
var sellingWood = function(m, n, prices) {
    const dp = Array(m + 1).fill(0).map(() => Array(n + 1).fill(0));

    for (const [h, w, p] of prices) {
        if (h <= m && w <= n) {
            dp[h][w] = Math.max(dp[h][w], p);
        }
    }

    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            // Consider horizontal cuts
            for (let k = 1; k < i; k++) {
                dp[i][j] = Math.max(dp[i][j], dp[k][j] + dp[i - k][j]);
            }
            // Consider vertical cuts
            for (let k = 1; k < j; k++) {
                dp[i][j] = Math.max(dp[i][j], dp[i][k] + dp[i][j - k]);
            }
        }
    }

    return dp[m][n];
};

File: MATTHEW/leetcode/AIcoding/1.Gemini-2.5/go_solutions/2312-selling-pieces-of-wood.go
package main

func sellingWood(m int, n int, prices [][]int) int64 {
    dp := make([][]int64, m+1)
    for i := range dp {
        dp[i] = make([]int64, n+1)
    }

    for _, price := range prices {
        r, c, p := price[0], price[1], price[2]
        if r <= m && c <= n {
            dp[r][c] = max(dp[r][c], int64(p))
        }
    }

    for i := 1; i <= m; i++ {
        for j := 1; j <= n; j++ {
            // Consider all possible horizontal cuts
            for k := 1; k < i; k++ {
                dp[i][j] = max(dp[i][j], dp[k][j]+dp[i-k][j])
            }
            // Consider all possible vertical cuts
            for k := 1; k < j; k++ {
                dp[i][j] = max(dp[i][j], dp[i][k]+dp[i][j-k])
            }
        }
    }

    return dp[m][n]
}

func max(a, b int64) int64 {
    if a > b {
        return a
    }
    return b
}